content/blog/2020/06/2020-06-03-git-performance-improvement-phase1.adoc

```diff
@@ -0,0 +1,82 @@
+---
+
+Git Plugin Performance Improvement is a Google Summer of Code 2020 project.
+It aims to improve the performance of the git plugin which provides fundamental git functionalities.
+
+Internally, the plugin provides these functionalities using two implementations: cli git and jgit(pure java implementation).
+
+
+image:/images/post-images/2020/07-git-plugin-performance-improvement/git-intro.png[git-intro]
+*cli git is the default implementation for the plugin, a user can switch to jgit if needed*
+
+The project is divided into two parallel stages:
+
+* *Stage 1*: Create benchmarks which evaluate the execution time of a git operation provided by cli git and jgit using JMH, which is a micro benchmarking test harness.
+* *Stage 2*: Implement the insights gained from the analysis into the plugin in order to improve the overall performance of the plugin.
+
+The project also aims to fix any existing performance bottlenecks within the plugin as well.
+
+
+== Benchmarks
+The benchmarks are written using link:https://openjdk.java.net/projects/code-tools/jmh/[JMH].
+
+* JMH is provided within the plugin through the Jenkins Unit Test Harness POM dependency.
+* The JMH link:https://github.com/jenkinsci/git-client-plugin/tree/master/src/test/java/jmh/benchmark[benchmarks] are created and run within the git client plugin
+* During phase-1, we have created benchmarks for two operations: "git fetch" and "git ls-remote"
+
+== Results and Analysis
+
+The benchmark analysis for git fetch: 
+
+=== Git fetch results
+
+image: /images/post-images/2020/07-git-plugin-performance-improvement/git-fetch-results.png[git-fetch-results]
+
+* The performance of git fetch (average execution time/op) is strongly correlated to the size of a repository
+* There exists an inflection point on the scale of repository size after which the nature of jgit performance changes (it starts to degrade)
+* After running multiple benchmarks, it is safe to say that for a large sized repository *cli-git* would be a better choice of implementation.
+* We can use this insight to implement a feature which avoids jgit when it comes to large repositories. 
+
+Please refer to link:https://github.com/jenkinsci/git-client-plugin/pull/521[PR-521] for an elaborate explantion on these results
+
+*Note:* Repository size means `du -h .git`
+
+== Fixing redundant fetch issue
+The git plugin performs two fetch operations instead of one while performing a fresh checkout of a remote git repository.
+
+To link:https://github.com/jenkinsci/git-plugin/pull/904[fix] this issue, we had to safely remove the second fetch keeping multiple use-cases in mind. The fix itself was not difficult to code but to do that safely without breaking any existing use-case was a challenging task.
+
+== Further Plan
+After consolidating a benchmarking strategy during Phase 1, the next steps will be: 
+
+* Provide a functionality to the git plugin which enables it to estimate the size of repository without cloning it.
```

Comment:
This sentence seems off `Provide a functionality to the git plugin which enables it to estimate the size of repository without cloning it.`
This sounds grammatically better `Provide functionality to the git plugin, which enables it to estimate the size of the repository without cloning it.`